extensions/azurecore/src/account-provider/azureAccountProvider.ts

```diff
@@ -144,17 +146,41 @@ export class AzureAccountProvider implements azdata.AccountProvider, vscode.Disp
 		await this.initCompletePromise;
 		const azureAuth = this.getAuthMethod(account);
 		if (azureAuth) {
+			let identifier = resource.toString() + '|' + tenantId;
 			Logger.piiSanitized(`Getting account security token for ${JSON.stringify(account.key)} (tenant ${tenantId}). Auth Method = ${azureAuth.userFriendlyName}`, [], []);
 			if (this.authLibrary === Constants.AuthLibrary.MSAL) {
+				// Fetch cached token from account properties if token is available and valid.
+				if (account.properties.tokenMap?.[identifier]) {
+					let encryptedAccessToken = account.properties.tokenMap[identifier];
+					try {
+						let accessToken = await this.msalCacheProvider.decryptToken(encryptedAccessToken!);
+						if (this.isValidToken(accessToken)) {
+							return accessToken;
+						}
+					} catch (e) {
+						// Log decrypt error and move on to fetching fresh access token.
+						Logger.info(`Decrypting access token failed with error: ${e}, fetching new access token.`);
+					}
+				}
 				tenantId = tenantId || account.properties.owningTenant.id;
 				let authResult = await azureAuth.getTokenMsal(account.key.accountId, resource, tenantId);
 				if (this.isAuthenticationResult(authResult) && authResult.account && authResult.account.idTokenClaims) {
 					const token: Token = {
 						key: authResult.account.homeAccountId,
 						token: authResult.accessToken,
 						tokenType: authResult.tokenType,
-						expiresOn: authResult.account.idTokenClaims.exp
+						expiresOn: authResult.account.idTokenClaims.exp!
```

Comment:
Not sure if you wanted to log anything here if we don't get an expiration time (which I imagine is not supposed to happen), but it wouldn't hurt. Overall though you should be careful with the non-null assertions, that's an easy way to have bugs slip in that are hard to track down. 

```suggestion
						expiresOn: authResult.account?.idTokenClaims?.exp
```

Reply:
It's already checked in the 'if' condition above: `&& authResult.account && authResult.account.idTokenClaims) `